pkg/trace/stats/oteltest/otel_apm_stats_comparison_test.go

```diff
@@ -0,0 +1,177 @@
+	"github.com/DataDog/datadog-agent/pkg/trace/stats"
+	"github.com/DataDog/opentelemetry-mapping-go/pkg/otlp/attributes"
+	"github.com/google/go-cmp/cmp"
+	"github.com/tj/assert"
+	"go.opentelemetry.io/collector/component/componenttest"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	semconv "go.opentelemetry.io/collector/semconv/v1.6.1"
+	"go.opentelemetry.io/otel/metric/noop"
+	"google.golang.org/protobuf/testing/protocmp"
+)
+
+// Comparison test to ensure APM stats generated from 2 different OTel ingestion paths are consistent.
+func TestOTelAPMStatsMatch(t *testing.T) {
+	ctx := context.Background()
+	set := componenttest.NewNopTelemetrySettings()
+	set.MeterProvider = noop.NewMeterProvider()
+	attributesTranslator, err := attributes.NewTranslator(set)
+	assert.NoError(t, err)
+	tcfg := getTraceAgentCfg(attributesTranslator)
+
+	// Set up 2 output channels for APM stats, and start 2 fake trace agent to conduct a comparison test
+	now := time.Now()
+	statschan1 := make(chan *pb.StatsPayload, 100)
+	fakeAgent1 := getAndStartFakeAgent(ctx, tcfg, statschan1, now)
+	defer fakeAgent1.Stop()
+	statschan2 := make(chan *pb.StatsPayload, 100)
+	fakeAgent2 := getAndStartFakeAgent(ctx, tcfg, statschan2, now)
+	defer fakeAgent2.Stop()
+
+	traces := getTestTraces()
+
+	// fakeAgent1 has OTLP traces go through the old pipeline: ReceiveResourceSpan -> TraceWriter -> ... ->  Concentrator.Run
+	fakeAgent1.Ingest(ctx, traces)
+
+	// fakeAgent2 calls the new API in Concentrator that directly calculates APM stats for OTLP traces
+	inputs := stats.OTLPTracesToConcentratorInputs(traces, tcfg, []string{semconv.AttributeContainerID, semconv.AttributeK8SContainerName})
+	for _, input := range inputs {
+		fakeAgent2.Concentrator.Add(input)
+	}
+
+	// Verify APM stats generated from 2 paths are consistent
+	var payload1 *pb.StatsPayload
+	var payload2 *pb.StatsPayload
+	for payload1 == nil || payload2 == nil {
+		select {
+		case sp1 := <-statschan1:
+			if len(sp1.Stats) > 0 {
+				payload1 = sp1
+				for _, csb := range sp1.Stats {
+					assert.Len(t, csb.Stats, 1)
+					assert.Len(t, csb.Stats[0].Stats, 3) // stats on 3 spans
+					sort.Slice(csb.Stats[0].Stats, func(i, j int) bool {
+						return csb.Stats[0].Stats[i].Name < csb.Stats[0].Stats[j].Name
+					})
+				}
+			}
+		case sp2 := <-statschan2:
+			if len(sp2.Stats) > 0 {
+				payload2 = sp2
+				for _, csb := range sp2.Stats {
+					assert.Len(t, csb.Stats, 1)
+					assert.Len(t, csb.Stats[0].Stats, 3) // stats on 3 spans
+					sort.Slice(csb.Stats[0].Stats, func(i, j int) bool {
+						return csb.Stats[0].Stats[i].Name < csb.Stats[0].Stats[j].Name
+					})
+				}
+			}
+		}
+	}
+
+	if diff := cmp.Diff(
+		payload1,
+		payload2,
+		protocmp.Transform(),
+		// OTLPTracesToConcentratorInputs adds container tags to ClientStatsPayload, other fields should match.
+		protocmp.IgnoreFields(&pb.ClientStatsPayload{}, "tags")); diff != "" {
```

Comment:
I'm guessing the check on container tags is happening below? Since it's only relevant to the payload2 case?

Reply:
That's right, payload1 (old path) doesn't have container tags, payload2 is validated below.